### DISCUSSION TITLE
Allow setting the timestamp of events submitted via the Events API

### DISCUSSION THREAD

**Discussion by rodrigotassinari** [plausible/analytics#3076] [Anything Else]
We're experimenting with Plausible at my company (an ecommerce website), and starting to use the [custom events](https://plausible.io/docs/custom-event-goals) feature (especially useful now with the recent addition of [revenue tracking](https://plausible.io/docs/ecommerce-revenue-tracking)).

At the [metrics definition docs](https://plausible.io/docs/metrics-definitions) there's a remark on the [bounce rate explanation](https://plausible.io/docs/metrics-definitions#bounce-rate) that "_all custom events are considered interactive_". So this means all custom events either extend the visitor's current visit or create a new one.

This is not an issue for most of the events that will be sent from our system to Plausible, either client-side or server-side, since they represent actions taken by the visitor during his active visit. However, we have a few crucial events that may happen hours or even days after the visitor's actions, such as our "Order Completed" event, which is fired only after payment confirmation, which depending on the payment method used, can happen long after the order was created (around 25% of our orders are paid in this asynchronous manner).

So if we send this "Order Completed" event to Plausible we will have an artificially higher number of visits and/or visit duration. If we don't send this event, we won't have precise revenue amount tracking on Plausible.

This could be solved if we could supply the date-time the event actually occured (in this case, the order's checkout date-time) when submitting the event to Plausible via the Events API, and Plausible used this date-time on visit calculations. I imagine this could be very helpful also in case of back-filling past events to Plausible, either custom events or pageviews, and would allow people to migrate to Plausible more easily, bringing data from previous systems to plausible using the API to back-fill the past data.

**Comment by samhibberd** [comment]
Backfilling historic data would be a game changer for us, allowing us to fully migrate to plausible.

**Comment by rodrigotassinari** [comment]
Another option to solve our specific problem of dealing with events happening after the user session would be to be able to mark some events as "non interactive" (similar to what Google Analytics allows). But I don't think it would be interesting for Plausible to do that, and it doesn't solve the use-case of needing to back-fill data. I've opened a s[eparate feature request](https://github.com/plausible/analytics/discussions/3105) for this though, to see if there's any interest going that route.
